docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,541 @@
+5. We also need to keep track of the model's `apply_fn`, the otpimizer's
+   `update_fn`, and the optimizer state, in addition to the data.
+
+
+If we put all this state into a dataclass then we only have to pass around a
+single object (and the train data):
+
+```python
+@flax.struct.dataclass
+class LinenHelper:
+  apply_fn: Callable = flax.struct.field(pytree_node=False)
+  variables: Dict[str, PyTree]
+  optim: Optimizer = flax.struct.field(pytree_node=False)
+  opt_state: OptimizerState
+```
+
+We can then augment this dataclass with some code (similar to
+`flax.optim.Optimizer` and `flax.nn.Model`) that takes care of splitting and
+merging the variable collections appropriately:
+
+```python
+  def update_with_loss(self, loss_fn, inputs, labels, *, mutable=False):
+
+    def loss_from_params(params):
+      outputs, new_model_state = self.apply_fn(
+        self.variables.copy(dict(params=params)), inputs, mutable=mutable or [])
+      loss = loss_fn(inputs, outputs, labels)
+      return loss, new_model_state
+
+    model_state, params = variables.pop('params')
+    (loss, new_model_state), grads = jax.value_and_grad(
+        loss_from_params, has_aux=True)(params)
+    new_params, new_opt_state = self.optim.update(grads, params, self.opt_state)
+
+    return self.replace(
+        variables=variables.copy(dict(params=new_params, **new_model_state)),
+        opt_state=new_opt_state,
+    ), loss
+```
+
+Users can derive from this dataclass and put any additional state into that
+object:
+
+```python
+@flax.struct.dataclass
+class TrainState(LinenHelper):
+  step: int
+```
+
+This will reduce the above training loop to something much simpler:
+
+```python
+@jax.jit
+def train_step(state, inputs, labels):
+
+  def loss_fn(inputs, outputs, labels):
+    del inputs  # unused
+    return xent_loss(outputs, labels)
+
+  state, loss = state.replace(step=state.step + 1).update_with_loss(
+      loss_fn, inputs, labels, mutable=['batch_stats'])
+  return state, loss
+
+opt_state = optim.init(variables['params'])
+state = TrainState(
+  apply_fn=model.apply,
+  variables=variables,
+  optim=optim,
+  opt_state=opt_state,
+  step=0,
+)
+for batch in ds.as_numpy_iterator():
+  state, loss = train_step(state, batch['image'], batch['label'])
+  print(loss)
+```
+
+Remarks:
+
+- It is quite common to have `TrainState`-like objects in Flax that keep all
+  the state (and some functions) and are used for checkpointing.
+- Variables can be inspected and additionally modified in `train_step()` if
+  needed.
+- `LinenHelper` might provide more utility functions with slightly different
+  signatures if there is demand for that.
+
+# Previous API
+[previous API]: #previous-api
+
+## Previous Implementation
+[Previous Implementation]: #previous-implementation
+
+The optimizer itself would be implemented by creating a new class derived
+from `OpimizerDef`:
+
+```python
+# flax/optim/momentum.py
+
+@flax.struct.dataclass
+class _MomentumHyperParams:
```

Comment:
Why did you remove two of the hyperparameters here? (`nesterov` and `weight_decay`). The reason I am asking is because I was confused since `nesterov` and `decay` are used in the "Using Optax" section, so I expect it to be here as well. Perhaps it is clearer to make both uses cases consistent?

Reply:
Note that the `optax.trace()` gradient transformation implements momentum; for simplicity I opted not to implement the `nesterov` in the `class Momentum(flax.optim.OptimizerDef)` example. Note that `optax` does not have a default argument for `trace(nesterov)` so I need to set it to `False` here.

Added a paragraph to the remarks to make this clearer.